src/web/rustdoc.rs

```diff
@@ -343,20 +338,44 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
     }
     let mut path = ctry!(req, percent_decode(path.as_bytes()).decode_utf8());
 
-    // Attempt to load the file from the database
-    let file = match File::from_path(&storage, &path, &config) {
-        Ok(file) => file,
-        Err(err) => {
-            log::debug!("got error serving {}: {}", path, err);
-            // If it fails, we try again with /index.html at the end
-            path.to_mut().push_str("/index.html");
-            req_path.push("index.html");
-
-            return if ctry!(req, storage.exists(&path)) {
-                redirect(&name, &version, &req_path[3..])
-            } else {
-                Err(Nope::ResourceNotFound.into())
-            };
+    let file = if krate.archive_storage {
+        let archive = format!("rustdoc/{0}/rustdoc-{0}-{1}.zip", name, version);
+        match File::from_archive_path(&storage, &archive, &path, &config) {
+            Ok(file) => file,
+            Err(err) => {
+                log::debug!("got error serving {}: {}", path, err);
+                // If it fails, we try again with /index.html at the end
+                path.to_mut().push_str("/index.html");
+                req_path.push("index.html");
+
+                return if ctry!(req, storage.exists_in_archive(&archive, &path)) {
+                    redirect(&name, &version, &req_path[3..])
+                } else {
+                    Err(Nope::ResourceNotFound.into())
+                };
+            }
+        }
+    } else {
```

Comment:
I'm not a giant fan of duplicating all this - maybe File could store whether it came from an archive internally? Then you could have `file.exists()` work whether it not it's in an archive, and remove almost all this code.

Reply:
as stated in the TODOs in the (long, I admin) description of the PR: 

I don't see this part here as finished, I also want a nicer way. I'll have to see how to restructure either the file or `web::rustdoc` for this.